File: ui/server/src/services/conversation-parser.ts
import { readdir, readFile } from "fs/promises";
import { join } from "path";

export interface TextContent {
  type: "text";
  text: string;
}

export interface ToolUseContent {
  type: "tool_use";
  id: string;
  name: string;
  input: Record<string, unknown>;
}

export interface ToolResultContent {
  type: "tool_result";
  tool_use_id: string;
  content: string;
}

export interface ConversationMessage {
  id: string;
  type: "assistant" | "user" | "system" | "iteration_start";
  timestamp?: string;
  content: (TextContent | ToolUseContent | ToolResultContent)[];
  toolResult?: unknown;
  toolUseId?: string;
  iteration?: number;
}

export interface ConversationTurn {
  assistant: ConversationMessage | null;
  toolCalls: {
    call: ToolUseContent;
    result: unknown;
  }[];
  iteration?: number;
}

export interface InitialPrompt {
  systemPrompt: string;
  userPrompt: string;
  timestamp?: string;
}

export interface IterationResult {
  iteration: number;
  subtype: string;
  durationMs: number;
  durationApiMs: number;
  numTurns: number;
  totalCostUsd: number;
  isError: boolean;
  usage: {
    inputTokens: number;
    outputTokens: number;
    cacheReadInputTokens: number;
    cacheCreationInputTokens: number;
  };
}

interface RawLogEntry {
  type: string;
  timestamp?: string;
  uuid?: string;
  message?: {
    role?: string;
    content?: Array<{
      type: string;
      text?: string;
      id?: string;
      name?: string;
      input?: Record<string, unknown>;
      tool_use_id?: string;
      content?: string;
    }>;
  };
  tool_use_result?: unknown;
  parent_tool_use_id?: string;
  session_id?: string;
  iteration?: number;
  system_prompt?: string;
  user_prompt?: string;
}

export interface ParseResult {
  messages: ConversationMessage[];
  initialPrompt: InitialPrompt | null;
  iterationResults: IterationResult[];
}

export async function parseIterationLogs(
  workerDir: string
): Promise<ParseResult> {
  const logsDir = join(workerDir, "logs");
  const messages: ConversationMessage[] = [];
  const iterationResults: IterationResult[] = [];
  let initialPrompt: InitialPrompt | null = null;

  try {
    const files = await readdir(logsDir);
    const iterationFiles = files
      .filter((f) => f.startsWith("iteration-") && f.endsWith(".log"))
      .sort((a, b) => {
        const numA = parseInt(a.match(/iteration-(\d+)/)?.[1] || "0");
        const numB = parseInt(b.match(/iteration-(\d+)/)?.[1] || "0");
        return numA - numB;
      });

    let currentIteration = 0;

    for (const file of iterationFiles) {
      const content = await readFile(join(logsDir, file), "utf-8");
      const lines = content.split("\n").filter((l) => l.trim());

      for (const line of lines) {
        try {
          const entry = JSON.parse(line);

          // Track iteration number
          if (entry.type === "iteration_start") {
            currentIteration = entry.iteration ?? currentIteration;

            // Capture initial prompt from first iteration_start
            if (!initialPrompt && (entry.system_prompt || entry.user_prompt)) {
              initialPrompt = {
                systemPrompt: entry.system_prompt || "",
                userPrompt: entry.user_prompt || "",
                timestamp: entry.timestamp,
              };
            }

            // Add iteration marker
            messages.push({
              id: `iter-start-${currentIteration}`,
              type: "iteration_start",
              timestamp: entry.timestamp,
              content: [{ type: "text", text: `Iteration ${currentIteration}` }],
              iteration: currentIteration,
            });
            continue;
          }

          // Capture iteration result/metrics
          if (entry.type === "result") {
            iterationResults.push({
              iteration: currentIteration,
              subtype: entry.subtype || "success",
              durationMs: entry.duration_ms || 0,
              durationApiMs: entry.duration_api_ms || 0,
              numTurns: entry.num_turns || 0,
              totalCostUsd: entry.total_cost_usd || 0,
              isError: entry.is_error || false,
              usage: {
                inputTokens: entry.usage?.input_tokens || 0,
                outputTokens: entry.usage?.output_tokens || 0,
                cacheReadInputTokens: entry.usage?.cache_read_input_tokens || 0,
                cacheCreationInputTokens: entry.usage?.cache_creation_input_tokens || 0,
              },
            });
            continue;
          }

          const msg = parseLogEntry(entry as RawLogEntry);
          if (msg) {
            msg.iteration = currentIteration;
            messages.push(msg);
          }
        } catch {
          // Skip invalid JSON lines
        }
      }
    }
  } catch {
    // Logs directory might not exist
  }

  return { messages, initialPrompt, iterationResults };
}

function parseLogEntry(entry: RawLogEntry): ConversationMessage | null {
  if (entry.type === "assistant" && entry.message?.content) {
    const content: (TextContent | ToolUseContent)[] = [];

    for (const item of entry.message.content) {
      if (item.type === "text" && item.text) {
        content.push({ type: "text", text: item.text });
      } else if (item.type === "tool_use" && item.id && item.name) {
        content.push({
          type: "tool_use",
          id: item.id,
          name: item.name,
          input: item.input || {},
        });
      }
    }

    if (content.length > 0) {
      return {
        id: entry.uuid || Math.random().toString(36),
        type: "assistant",
        timestamp: entry.timestamp,
        content,
      };
    }
  }

  if (entry.type === "user" && entry.message?.content) {
    const firstItem = entry.message.content[0];

    // Check if this is a tool result
    if (firstItem?.type === "tool_result" && firstItem.tool_use_id) {
      return {
        id: entry.uuid || Math.random().toString(36),
        type: "user",
        timestamp: entry.timestamp,
        content: [],
        toolResult: entry.tool_use_result,
        toolUseId: firstItem.tool_use_id,
      };
    }
  }

  return null;
}

export function groupIntoTurns(
  messages: ConversationMessage[]
): ConversationTurn[] {
  const turns: ConversationTurn[] = [];
  let currentTurn: ConversationTurn | null = null;
  let currentIteration = 0;
  const pendingToolCalls = new Map<string, ToolUseContent>();

  for (const msg of messages) {
    // Handle iteration boundaries
    if (msg.type === "iteration_start") {
      // Save previous turn if exists
      if (currentTurn && (currentTurn.assistant || currentTurn.toolCalls.length > 0)) {
        turns.push(currentTurn);
      }
      currentIteration = msg.iteration ?? 0;
      currentTurn = null;
      pendingToolCalls.clear();
      continue;
    }

    if (msg.type === "assistant") {
      // Check if this message has text content (start new turn)
      const hasText = msg.content.some(
        (c) => c.type === "text" && (c as TextContent).text.trim()
      );

      if (hasText) {
        // Save previous turn if exists
        if (currentTurn && (currentTurn.assistant || currentTurn.toolCalls.length > 0)) {
          turns.push(currentTurn);
        }
        currentTurn = { assistant: msg, toolCalls: [], iteration: currentIteration };
      } else if (!currentTurn) {
        currentTurn = { assistant: null, toolCalls: [], iteration: currentIteration };
      }

      // Collect tool calls
      for (const item of msg.content) {
        if (item.type === "tool_use") {
          pendingToolCalls.set((item as ToolUseContent).id, item as ToolUseContent);
        }
      }
    } else if (msg.type === "user" && msg.toolUseId && msg.toolResult !== undefined) {
      const toolCall = pendingToolCalls.get(msg.toolUseId);
      if (toolCall) {
        if (!currentTurn) {
          currentTurn = { assistant: null, toolCalls: [], iteration: currentIteration };
        }
        currentTurn.toolCalls.push({
          call: toolCall,
          result: msg.toolResult,
        });
        pendingToolCalls.delete(msg.toolUseId);
      }
    }
  }

  // Push final turn
  if (currentTurn && (currentTurn.assistant || currentTurn.toolCalls.length > 0)) {
    turns.push(currentTurn);
  }

  return turns;
}

File: tui/wiggum_tui/data/conversation_parser.py
"""Conversation parser for iteration log files."""

import json
from pathlib import Path
from typing import Any
from .models import (
    Conversation,
    ConversationTurn,
    ToolCall,
    IterationResult,
    TokenUsage,
)


# Cache for parsed conversations: worker_path -> (max_mtime, conversation)
_conversation_cache: dict[str, tuple[float, Conversation]] = {}


def get_logs_max_mtime(logs_dir: Path) -> float:
    """Get the maximum modification time of all log files.

    Args:
        logs_dir: Path to logs directory.

    Returns:
        Maximum mtime of any log file, or 0 if no logs found.
    """
    if not logs_dir.is_dir():
        return 0.0
    max_mtime = 0.0
    try:
        for log_file in logs_dir.glob("**/*.log"):
            try:
                mtime = log_file.stat().st_mtime
                if mtime > max_mtime:
                    max_mtime = mtime
            except OSError:
                continue
    except OSError:
        pass
    return max_mtime


def clear_conversation_cache(worker_id: str | None = None) -> None:
    """Clear the conversation cache.

    Args:
        worker_id: If provided, only clear cache for this worker.
            If None, clear entire cache.
    """
    global _conversation_cache
    if worker_id:
        _conversation_cache.pop(worker_id, None)
    else:
        _conversation_cache.clear()


def has_logs_changed(worker_dir: Path) -> bool:
    """Check if logs have changed since last parse.

    Args:
        worker_dir: Path to worker directory.

    Returns:
        True if logs have changed or not cached, False if unchanged.
    """
    logs_dir = worker_dir / "logs"
    worker_id = worker_dir.name

    if not logs_dir.is_dir():
        return False

    current_mtime = get_logs_max_mtime(logs_dir)
    if worker_id in _conversation_cache:
        cached_mtime, _ = _conversation_cache[worker_id]
        return current_mtime > cached_mtime
    return True  # Not cached, consider it changed


def parse_iteration_logs(worker_dir: Path, use_cache: bool = True) -> Conversation:
    """Parse all iteration logs for a worker.

    Args:
        worker_dir: Path to worker directory.
        use_cache: If True, use cached result if logs haven't changed.

    Returns:
        Conversation object with all turns and results.
    """
    global _conversation_cache

    logs_dir = worker_dir / "logs"
    worker_id = worker_dir.name

    if not logs_dir.is_dir():
        return Conversation(worker_id=worker_id)

    # Check cache
    if use_cache:
        current_mtime = get_logs_max_mtime(logs_dir)
        if worker_id in _conversation_cache:
            cached_mtime, cached_conv = _conversation_cache[worker_id]
            if cached_mtime >= current_mtime:
                return cached_conv

    conversation = Conversation(worker_id=worker_id)

    # Sort base logs by mtime, then insert -summary logs directly after their base.
    # A -summary file is only paired if a matching base log exists; otherwise it's
    # treated as a regular log and sorted by its own mtime.
    # Use **/*.log to find logs in nested subdirectories (e.g., logs/execution-123/...)
    all_logs = list(logs_dir.glob("**/*.log"))
    base_stems = {f.stem for f in all_logs if not f.stem.endswith("-summary")}
    summary_by_base: dict[str, Path] = {}
    non_summary: list[Path] = []
    for f in all_logs:
        if f.stem.endswith("-summary"):
            base_name = f.stem.removesuffix("-summary")
            if base_name in base_stems:
                summary_by_base[base_name] = f
            else:
                non_summary.append(f)
        else:
            non_summary.append(f)
    non_summary.sort(key=lambda f: f.stat().st_mtime)
    log_files: list[Path] = []
    for base in non_summary:
        log_files.append(base)
        if base.stem in summary_by_base:
            log_files.append(summary_by_base[base.stem])

    all_entries: list[dict[str, Any]] = []
    results: list[IterationResult] = []

    for idx, log_file in enumerate(log_files):
        log_name = log_file.stem  # e.g., "iteration-0", "validation-review"
        try:
            content = log_file.read_text()
            for line in content.split("\n"):
                if not line.strip():
                    continue
                try:
                    entry = json.loads(line)
                    entry_type = entry.get("type")
                    # Tag entry with its iteration index and log name
                    entry["_iteration_idx"] = idx
                    entry["_log_name"] = log_name

                    if entry_type == "iteration_start":
                        if not conversation.system_prompt:
                            conversation.system_prompt = entry.get("system_prompt", "")
                            conversation.user_prompt = entry.get("user_prompt", "")
                    elif entry_type == "result":
                        result = _parse_result(entry)
                        result.iteration = idx  # Override with mtime-based index
                        result.log_name = log_name
                        results.append(result)
                    elif entry_type in ("assistant", "user"):
                        all_entries.append(entry)
                except json.JSONDecodeError:
                    continue
        except OSError:
            continue

    # Group messages into turns
    conversation.turns = _group_into_turns(all_entries)
    conversation.results = results

    # Cache the parsed conversation
    if use_cache:
        current_mtime = get_logs_max_mtime(logs_dir)
        _conversation_cache[worker_id] = (current_mtime, conversation)

    return conversation


def _parse_result(entry: dict[str, Any]) -> IterationResult:
    """Parse a result entry into IterationResult."""
    usage_data = entry.get("usage", {})
    usage = TokenUsage(
        input=usage_data.get("input_tokens", 0),
        output=usage_data.get("output_tokens", 0),
        cache_creation=usage_data.get("cache_creation_input_tokens", 0),
        cache_read=usage_data.get("cache_read_input_tokens", 0),
    )
    usage.total = usage.input + usage.output

    return IterationResult(
        iteration=entry.get("iteration", 0),
        subtype=entry.get("subtype", ""),
        duration_ms=entry.get("duration_ms", 0),
        duration_api_ms=entry.get("duration_api_ms", 0),
        num_turns=entry.get("num_turns", 0),
        total_cost_usd=entry.get("total_cost_usd", 0.0),
        is_error=entry.get("is_error", False),
        usage=usage,
    )


def _group_into_turns(entries: list[dict[str, Any]]) -> list[ConversationTurn]:
    """Group messages into conversation turns.

    A turn consists of an assistant message (text and/or tool calls) followed by tool results.
    """
    turns: list[ConversationTurn] = []

    # Track all tool calls by ID for result linking
    all_tool_calls: dict[str, ToolCall] = {}
    # Track pending tool calls that need results (in order)
    pending_tool_ids: list[str] = []

    current_text: str | None = None
    current_tool_calls: list[ToolCall] = []
    current_timestamp: str | None = None
    current_iteration = 0
    current_log_name = ""

    def save_turn():
        nonlocal current_text, current_tool_calls, current_timestamp
        if current_text or current_tool_calls:
            turns.append(ConversationTurn(
                iteration=current_iteration,
                assistant_text=current_text,
                tool_calls=current_tool_calls.copy(),
                timestamp=current_timestamp,
                log_name=current_log_name,
            ))
        current_text = None
        current_tool_calls = []
        current_timestamp = None

    for entry in entries:
        entry_type = entry.get("type")
        entry_iteration = entry.get("_iteration_idx", 0)
        entry_log_name = entry.get("_log_name", "")

        # Skip nested messages from subagents (Task tool execution)
        # These have parent_tool_use_id set to the Task tool's ID
        if entry.get("parent_tool_use_id"):
            continue

        if entry_type == "assistant":
            current_iteration = entry_iteration
            current_log_name = entry_log_name
            message_content = entry.get("message", {}).get("content", [])

            for block in message_content:
                block_type = block.get("type")

                if block_type == "text":
                    # Text block - save previous turn and start new one
                    save_turn()
                    current_text = block.get("text", "")
                    current_timestamp = entry.get("timestamp")

                elif block_type == "tool_use":
                    # Tool use block - add to current turn
                    tool_id = block.get("id", "")
                    tool_call = ToolCall(
                        name=block.get("name", ""),
                        input=block.get("input", {}),
                    )
                    all_tool_calls[tool_id] = tool_call
                    pending_tool_ids.append(tool_id)
                    current_tool_calls.append(tool_call)
                    if not current_timestamp:
                        current_timestamp = entry.get("timestamp")

        elif entry_type == "user":
            # Tool result
            result = entry.get("tool_use_result")
            parent_id = entry.get("parent_tool_use_id")

            # If parent_tool_use_id is not set, try to extract from message.content
            if not parent_id:
                message_content = entry.get("message", {}).get("content", [])
                for block in message_content:
                    if block.get("type") == "tool_result":
                        parent_id = block.get("tool_use_id")
                        break

            if parent_id and parent_id in all_tool_calls:
                # Explicit parent ID - link to that tool call
                all_tool_calls[parent_id].result = result
                if parent_id in pending_tool_ids:
                    pending_tool_ids.remove(parent_id)
            elif pending_tool_ids:
                # No parent ID - link to first pending tool call
                first_pending = pending_tool_ids.pop(0)
                all_tool_calls[first_pending].result = result

    # Save final turn
    save_turn()

    return turns


def get_conversation_summary(conversation: Conversation) -> dict[str, Any]:
    """Get summary statistics for a conversation.

    Args:
        conversation: Conversation to summarize.

    Returns:
        Dictionary with summary statistics.
    """
    total_turns = len(conversation.turns)
    total_tool_calls = sum(len(t.tool_calls) for t in conversation.turns)

    total_cost = sum(r.total_cost_usd for r in conversation.results)
    total_duration_ms = sum(r.duration_ms for r in conversation.results)
    total_tokens = sum(
        r.usage.input + r.usage.output for r in conversation.results
    )

    return {
        "turns": total_turns,
        "tool_calls": total_tool_calls,
        "cost_usd": total_cost,
        "duration_ms": total_duration_ms,
        "tokens": total_tokens,
        "iterations": len(conversation.results),
    }


def truncate_text(text: str, max_length: int = 100) -> str:
    """Truncate text for display.

    Args:
        text: Text to truncate.
        max_length: Maximum length.

    Returns:
        Truncated text with ellipsis if needed.
    """
    if len(text) <= max_length:
        return text
    return text[: max_length - 3] + "..."


def format_tool_result(result: Any, max_length: int = 200) -> str:
    """Format a tool result for display.

    Args:
        result: The tool result (can be dict, str, or None).
        max_length: Maximum length of output.

    Returns:
        Formatted string representation.
    """
    if result is None:
        return "(no result)"

    if isinstance(result, dict):
        # Handle common result types
        if "type" in result:
            result_type = result.get("type")
            if result_type == "text":
                # File read result
                file_info = result.get("file", {})
                file_path = file_info.get("filePath", "")
                if file_path:
                    return f"Read: {file_path}"
            elif result_type == "tool_result":
                content = result.get("content", "")
                if isinstance(content, str):
                    return truncate_text(content, max_length)

        if "stdout" in result:
            return truncate_text(result["stdout"], max_length)
        if "error" in result:
            return f"Error: {truncate_text(str(result['error']), max_length)}"
        if "success" in result:
            return "Success" if result["success"] else "Failed"

        # Generic dict - show keys
        keys = list(result.keys())[:5]
        return f"{{...}} ({', '.join(keys)})"

    return truncate_text(str(result), max_length)
